gammapy/maps/geom.py

```diff
@@ -631,29 +631,79 @@ def energy_mask(self, energy_min=None, energy_max=None, round_to_edge=False):
         ----------
         energy_min, energy_max : `~astropy.units.Quantity`
             Energy range.
+        round_to_edge: bool, optional
+            Wether to round `energy_min` and `energy_max` to the closest axis bin value.
+            See `~gammapy.maps.MapAxis.round`. Default is False.
 
         Returns
         -------
         mask : `~gammapy.maps.Map`
             Map containing the energy mask. The geometry of the map
             is the same as the geometry of the instance which called this method.
         """
-        from . import Map
+        return self.create_mask(
+            "energy",
+            edge_min=energy_min,
+            edge_max=energy_max,
+            round_to_edge=round_to_edge,
+        )
 
-        # get energy axes and values
-        energy_axis = self.axes["energy"]
+    def create_mask(self, axis_name, edge_min=None, edge_max=None, round_to_edge=False):
+        """Create a mask over a given axis.
 
-        if round_to_edge:
-            energy_min, energy_max = energy_axis.round([energy_min, energy_max])
+        The energy bin must be fully contained to be included in the mask.
 
-        # TODO: make this more general
-        shape = (-1, 1) if self.is_hpx else (-1, 1, 1)
-        energy_edges = energy_axis.edges.reshape(shape)
+        Parameters
+        ----------
+        edge_min, edge_max : `~astropy.units.Quantity` or float
+            Bounds of the mask. If the axis used for the mask has a unit, passing `~astropy.units.Quantity` is mandatory.
+        round_to_edge: bool, optional
+            Wether to round `energy_min` and `energy_max` to the closest axis bin value.
+            See `~gammapy.maps.MapAxis.round`. Default is False.
 
-        # set default values
-        energy_min = energy_min if energy_min is not None else energy_edges[0]
-        energy_max = energy_max if energy_max is not None else energy_edges[-1]
+        Returns
+        -------
+        mask : `~gammapy.maps.Map`
+            Map containing the mask. The geometry of the map is the same as the initial geometry.
+        """
+        from . import Map
+
+        axes_names = self.axes_names
+
+        try:
+            axis = self.axes[axis_name]
+        except KeyError:
+            raise ValueError(
+                f"Axis with name `{axis_name}` not in list of axis name: {axes_names}."
+            )
+        shape = tuple([-1] + [1 for n in axes_names if n != axis_name])
+        axis_edges = axis.edges.reshape(shape)
+        edge_min = edge_min if edge_min is not None else axis_edges[0]
+        edge_max = edge_max if edge_max is not None else axis_edges[-1]
+
+        if axis.unit != "":
```

Comment:
I think this check is not really need. It will fail later in the comparison operator with a similar message:
```python
from astropy import units as u

edges = [1, 2, 3, 4] * u.TeV
edges > 1 * u.m
```
```python
UnitConversionError: Can only apply 'greater' function to quantities with compatible dimensions
```
If you feel strongly about the improved error message, maybe introduce a little helper function in `gammapy.utils` like `check_unit`.